File: 234. Palindrome Linked List.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        int count = 0;
        ListNode* node = head;

        while(node){
            ++count;
            node = node->next;
        }

        count /= 2;
        node = head;

        while(count--){
            node = node->next;
        }

        ListNode* before = nullptr, * after = nullptr;
        while(node){
            after = node->next;
            node->next = before;
            before = node;
            node = after;
        }
        node = before;

        while(node && head){
            if(node->val != head->val) return false;
            node = node->next;
            head = head->next;
        }

        return true;
    }
};

File: 234. Palindrome Linked List.java
public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public boolean isPalindrome(ListNode head) {
        int count = 0;
        ListNode node = head;
        while(node != null){
            ++count;
            node = node.next;
        }

        count /= 2;
        node = head;

        while(count > 0){
            --count;
            node = node.next;
        }

        ListNode after = null, before = null;
        while(node != null){
            after = node.next;
            node.next = before;
            before = node.next;
            node = after;
        }
        node = before;

        while(node != null && head != null){
            if(node.val != head.val) return false;
            node = node.next;
            head = head.next;
        }

        return true;
    }
}